Feature/issue 2 historical prices

jamie-munro opened the pull request:
Changed date representations to java.time.LocalDate
Added support for IEX historical range feature
Added validation to dates
Added validation to ranges
Removed unnecessary lists
Added tests for range feature, and for invalid dates/ranges
![Capture](https://user-images.githubusercontent.com/115724737/196750337-a865a620-94e6-412b-9d52-dd7028f6637d.PNG)
All tests passing

jmunro-verify-finance committed:
New FUSE API endpoint: /iex/historicalPrice.
New test defined to test the historicalPrice endpoint with valid parameters.
New tests defined to test the historicalPrice endpoint with invalid parameters.
Wiremock simulation of IEX historical-prices added.
All tests passing.

jmunro-verify-finance committed:
Changed date representations to java.time.LocalDate
Added support for IEX historical range feature
Added validation to dates
Added validation to ranges
Removed unnecessary lists
Added tests for range feature, and for invalid dates/ranges
All tests passing

GalateaFarhaad commented on src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java:
```diff
@@ -0,0 +1,20 @@
+public class IexHistoricalPrice {
+
+  private String symbol;
+  private Date date;
```
Use LocalDate here as well

jamie-munro replied:
@GalateaFarhaad  Had to use java.time.Instant instead of LocalDate as Spring doesn't know how to convert between UNIX timestamps and LocalDate without a timezone being specified. Hopefully this is ok as it is still a Java 8 class and I think it better represents the information returned by the API (which is unzoned)

GalateaFarhaad commented on src/main/java/org/galatea/starter/entrypoint/IexRestController.java:
```diff
@@ -42,10 +52,73 @@ public List<IexSymbol> getAllStockSymbols() {
+        return iexService.getHistoricalPriceOnDate(symbols.get(0), dates.get(0));
+      }
+      else {
+        return Collections.emptyList();
```
Generally, it's better to return an error if something happens. Returning an empty response is ambiguous for the user, because it's hard to tell whether the empty response is due to no data or due to an error.

There may be some cases in which we want to do it anyway, due to certain requirements we might have. But for this endpoint let's say we do want to return an error. The easiest way to do that is to just throw an exception. There are ways we can customize the response associated with certain exceptions, so try to do that here - we want a 400 Bad Request response code here, rather than a 500 Internal Server Error response (which is the default status code for most exceptions).

GalateaFarhaad commented on src/main/java/org/galatea/starter/entrypoint/IexRestController.java:
```diff
@@ -42,10 +52,73 @@ public List<IexSymbol> getAllStockSymbols() {
+  public List<IexHistoricalPrice> getHistoricalPrice(
+          @RequestParam(value = "symbol", required = true) final List<String> symbols,
+          @RequestParam(value = "range", required = false) final List<String> ranges,
+          @RequestParam(value = "date", required = false) @DateTimeFormat(pattern = "yyyyMMdd") final List<LocalDate> dates) {
```
These parameters don't need to be lists, since we only ever use the first item in each list (if one even exists). We can just use String/LocalDate.

jamie-munro committed:
Date field for IexHistoricalPrice changed to type java.time.Instant (LocalDate could not be used as Spring cannot convert directly between UNIX timestamp and LocalDate as the timezone is not specified)
Unnecessary lists in IexRestController removed
Error responses returned instead of empty lists when invalid input supplied for historicalPrices API route.
